### PR TITLE
CDAP-1691: In CLI debug mode, log HTTP requests and responses

### PR DESCRIPTION
**Pull request by jwang47**
https://issues.cask.co/browse/CDAP-1691

### CODE REVIEW

**Comment by jwang47** [comment]
http://builds.cask.co/browse/CDAP-DUT974

File: cdap-cli/src/main/java/co/cask/cdap/cli/CLIMain.java [hunk 1]
```diff
@@ -114,7 +119,38 @@ public CLIMain(PrintStream output,
...
+          output.printf("DEBUG: Executing HTTP request (attempt %d): " +
+                        "method=%s url=%s hasBody=%s bodyLength=%s headers=%s",
+                        attempt, request.getMethod().name(), request.getURL().toString(),
+                        request.getBody() != null, request.getBodyLength(), GSON.toJson(request.getHeaders()));
+          output.println();
+        }
+
+        @Override
+        public void onResponse(HttpRequest request, HttpResponse response, int attemptsMade) {
+          output.printf("DEBUG: Got HTTP response (after %d attempt(s)): " +
```
Comment (anew) [inline cdap-cli/src/main/java/co/cask/cdap/cli/CLIMain.java]: do you need to log the request again? Seems overly verbose...

Comment (jwang47) [inline cdap-cli/src/main/java/co/cask/cdap/cli/CLIMain.java]: True, I'll not log the request in `onResponse`.

**Comment by jwang47** [comment]
Build passed http://builds.cask.co/browse/CDAP-DUT974

File: cdap-client/src/main/java/co/cask/cdap/client/StreamClient.java [hunk 2]
```diff
@@ -100,7 +105,7 @@ public StreamProperties getConfig(String streamId) throws IOException, StreamNot
    * @throws IOException if a network error occurred
    * @throws UnauthorizedException if the client is unauthorized
    * @throws BadRequestException if the request is bad
-   * @throws StreamNotFoundException if the stream was not found
+   * @throws StreamNotFoundException if the stream was not stream was not found
```
Comment (anew) [inline cdap-client/src/main/java/co/cask/cdap/client/StreamClient.java]: looks like that happened by accident?

File: cdap-client/src/main/java/co/cask/cdap/client/util/RESTClient.java [hunk 3]
```diff
@@ -39,34 +42,32 @@
...
 
+  private final List<Listener> listeners;
   private final HttpRequestConfig defaultConfig;
   private final HttpRequestConfig uploadConfig;
   private final int unavailableRetryLimit;
 
-  public RESTClient(HttpRequestConfig defaultConfig, HttpRequestConfig uploadConfig, int retryLimit) {
+  @Inject
+  public RESTClient(ClientConfig clientConfig) {
+    this.listeners = Lists.newArrayList();
```
Comment (anew) [inline cdap-client/src/main/java/co/cask/cdap/client/util/RESTClient.java]: better call this(clientConfig.getDefaultHttpConfig(), clientConfig.getUploadHttpConfig(), clientConfig.getUnavailableRetryLimit());

**Comment by anew** [comment]
I have only minor comments, and one question:
Is it possible to turn on/off debug mode when the CLI is running, or do I have to exit and start a new CLI?

**Comment by jwang47** [comment]
It's not possible right now. Created an issue for that: https://issues.cask.co/browse/CDAP-1703

**Pull request merged**